elliptic-curve/src/group_digest.rs

```diff
@@ -0,0 +1,86 @@
+use crate::hash2field::{hash_to_field, ExpandMsg, FromOkm};
+use crate::isogeny::Isogeny;
+use crate::osswu::OsswuMap;
+use ff::Field;
+use group::Group;
+
+/// Convert a set of field elements into a point
+pub trait ConvertToPoint {
+    /// The field element that represents a point
+    type FieldElement: Field;
+    /// The result of converting two field elements to a Point
+    type Output;
+
+    /// Converts two coordinates into a point
+    fn convert_to_point(x: Self::FieldElement, y: Self::FieldElement) -> Self::Output;
+}
+
+/// Method for points to be cleared by the curve cofactor
+pub trait ClearCofactor {
+    /// Multiply this point by the curve cofactor to ensure
+    /// correct subgroup membership
+    fn clear_cofactor(&mut self);
+}
+
+/// Adds hashing arbitrary byte sequences to a valid group element
+pub trait GroupDigest {
+    /// The field element representation for a group value with multiple elements
+    type FieldElement: FromOkm + OsswuMap + Isogeny;
+    /// The resulting group element
+    type Output: Group
+        + ClearCofactor
+        + ConvertToPoint<FieldElement = Self::FieldElement, Output = Self::Output>;
+
+    /// Computes the hash to curve routine according to
+    /// <https://www.ietf.org/archive/id/draft-irtf-cfrg-hash-to-curve-13.html>
+    /// which says
+    /// Uniform encoding from byte strings to points in G.
+    /// That is, the distribution of its output is statistically close
+    /// to uniform in G.
+    /// This function is suitable for most applications requiring a random
+    /// oracle returning points in G assuming a cryptographically secure
+    /// hash function is used.
+    ///
+    /// Example using a fixed size hash function
+    ///
+    /// let pt = ProjectivePoint::hash_from_bytes::<hash2field::ExpandMsgXmd<sha2::Sha256>>(b"test data", b"CURVE_XMD:SHA-256_SSWU_RO_");
+    ///
+    ///
+    /// Example using an extendable output function
+    ///
+    /// let pt = ProjectivePoint::hash_from_bytes::<hash2field::ExpandMsgXof<sha3::Shake256>>(b"test data", b"CURVE_XOF:SHAKE-256_SSWU_RO_");
+    ///
+    fn hash_from_bytes<X: ExpandMsg, M: AsRef<[u8]>>(msg: M, dst: &'static [u8]) -> Self::Output {
+        let mut u = [Self::FieldElement::default(); 2];
+        hash_to_field::<X, _>(msg.as_ref(), dst, &mut u);
+        let (q0x, q0y) = u[0].osswu();
+        let (r0x, r0y) = Self::FieldElement::isogeny(q0x, q0y);
+        let (q1x, q1y) = u[1].osswu();
+        let (r1x, r1y) = Self::FieldElement::isogeny(q1x, q1y);
+        let mut q0 = Self::Output::convert_to_point(r0x, r0y);
+        let mut q1 = Self::Output::convert_to_point(r1x, r1y);
+        // Ideally we could add and then clear cofactor thus saving a multiplication
+        // but the field elements may not reduced by the isogeny modulus
+        // which would result in an incorrect subgroup
+        q0.clear_cofactor();
+        q1.clear_cofactor();
+        q0 + q1
```

Comment:
I think the spec should be followed more closely here:
https://www.ietf.org/archive/id/draft-irtf-cfrg-hash-to-curve-13.html#section-3-7
```
hash_to_curve(msg)

Input: msg, an arbitrary-length byte string.
Output: P, a point in G.

Steps:
1. u = hash_to_field(msg, 2)
2. Q0 = map_to_curve(u[0])
3. Q1 = map_to_curve(u[1])
4. R = Q0 + Q1              # Point addition
5. P = clear_cofactor(R)
6. return P
```
Again, I think a `MapToCurve` trait would fill the gap. OSSWU isn't applicable to all kind of groups. I understand that currently this is targeted a P-256 & co., but these traits will be used by other groups too.

Reply:
Yep will do.